File: flutter-external/lib/services/identity_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../config.dart';
import '../models/api_models.dart';
import '../models/session_context.dart';
import '../models/step_result.dart';
import 'crypto_service.dart';
import 'hmac_service.dart';

/// Token operations with AES-256-GCM encryption.
/// External client: HMAC-SHA256 for token issue, Bearer for steps 3/5/6.
class IdentityService {
  final CryptoService _crypto;
  final HmacService _hmac = HmacService();

  IdentityService(this._crypto);

  /// Step 2: Issue tokens (HMAC-SHA256 + GCM).
  /// HMAC is computed over the PLAINTEXT body (before encryption).
  Future<StepResult> issueToken(SessionContext session, String plaintext) async {
    final sw = Stopwatch()..start();
    try {
      final nonce = _crypto.generateNonce();
      final ts = DateTime.now().millisecondsSinceEpoch.toString();
      final aad = _crypto.buildAad(ts, nonce, session.kid, session.clientId);

      // Compute HMAC over plaintext body BEFORE encryption
      final signature = await _hmac.computeSignature(
        'POST',
        '/api/v1/token/issue',
        ts,
        nonce,
        plaintext,
        AppConfig.clientSecret,
      );

      final encryptedPayload = await _crypto.encrypt(
        plaintext,
        session.sessionKey,
        aad,
      );

      // Build request headers (HMAC instead of Basic Auth)
      final headers = <String, String>{
        'Content-Type': 'application/json',
        'X-Kid': session.kid,
        'X-ClientId': session.clientId,
        'X-Idempotency-Key': '$ts.$nonce',
        'X-Signature': signature,
      };

      final requestBody = jsonEncode(
        EncryptedPayload(payload: encryptedPayload).toJson(),
      );

      final response = await http.post(
        Uri.parse('${AppConfig.sidecarUrl}/api/v1/token/issue'),
        headers: headers,
        body: requestBody,
      );

      final respHeaders = _extractResponseHeaders(response);

      if (response.statusCode != 200) {
        sw.stop();
        return StepResult(
          step: 2,
          name: 'Token Issue (HMAC-SHA256 + GCM)',
          requestHeaders: headers,
          requestBodyPlaintext: plaintext,
          requestBodyEncrypted: encryptedPayload,
          responseHeaders: respHeaders,
          responseBodyDecrypted: response.body,
          durationMs: sw.elapsedMilliseconds,
          error: 'HTTP ${response.statusCode}: ${response.body}',
        );
      }

      final (respEncrypted, decrypted) = await _decryptResponse(
        response.body,
        respHeaders,
        session,
      );

      sw.stop();
      return StepResult(
        step: 2,
        name: 'Token Issue (HMAC-SHA256 + GCM)',
        requestHeaders: headers,
        requestBodyPlaintext: plaintext,
        requestBodyEncrypted: encryptedPayload,
        responseHeaders: respHeaders,
        responseBodyEncrypted: respEncrypted,
        responseBodyDecrypted: decrypted,
        durationMs: sw.elapsedMilliseconds,
        success: true,
      );
    } catch (e) {
      sw.stop();
      return StepResult(
        step: 2,
        name: 'Token Issue (HMAC-SHA256 + GCM)',
        requestBodyPlaintext: plaintext,
        durationMs: sw.elapsedMilliseconds,
        error: e.toString(),
      );
    }
  }

  /// Step 3: Introspect token (Bearer + GCM).
  Future<StepResult> introspectToken(
    SessionContext session,
    String plaintext,
    String accessToken,
  ) {
    return _postEncrypted(
      '/v1/introspect',
      session,
      plaintext,
      {'Authorization': 'Bearer $accessToken'},
      3,
      'Token Introspection (Bearer + GCM)',
    );
  }

  /// Step 5: Refresh tokens (Bearer + GCM).
  Future<StepResult> refreshToken(
    SessionContext session,
    String plaintext,
    String accessToken,
  ) {
    return _postEncrypted(
      '/v1/token',
      session,
      plaintext,
      {'Authorization': 'Bearer $accessToken'},
      5,
      'Token Refresh (Bearer + GCM)',
    );
  }

  /// Step 6: Revoke token (Bearer + GCM).
  Future<StepResult> revokeToken(
    SessionContext session,
    String plaintext,
    String accessToken,
  ) {
    return _postEncrypted(
      '/v1/revoke',
      session,
      plaintext,
      {'Authorization': 'Bearer $accessToken'},
      6,
      'Token Revocation (Bearer + GCM)',
    );
  }

  /// Core: encrypt body -> POST -> decrypt response.
  Future<StepResult> _postEncrypted(
    String path,
    SessionContext session,
    String plaintext,
    Map<String, String> authHeaders,
    int stepNum,
    String stepName,
  ) async {
    final sw = Stopwatch()..start();
    try {
      final nonce = _crypto.generateNonce();
      final ts = DateTime.now().millisecondsSinceEpoch.toString();
      final aad = _crypto.buildAad(ts, nonce, session.kid, session.clientId);

      final encryptedPayload = await _crypto.encrypt(
        plaintext,
        session.sessionKey,
        aad,
      );

      final headers = <String, String>{
        'Content-Type': 'application/json',
        'X-Kid': session.kid,
        'X-Idempotency-Key': '$ts.$nonce',
        'X-ClientId': session.clientId,
        ...authHeaders,
      };

      final requestBody = jsonEncode(
        EncryptedPayload(payload: encryptedPayload).toJson(),
      );

      final response = await http.post(
        Uri.parse('${AppConfig.sidecarUrl}/api$path'),
        headers: headers,
        body: requestBody,
      );

      final respHeaders = _extractResponseHeaders(response);

      if (response.statusCode != 200) {
        sw.stop();
        return StepResult(
          step: stepNum,
          name: stepName,
          requestHeaders: headers,
          requestBodyPlaintext: plaintext,
          requestBodyEncrypted: encryptedPayload,
          responseHeaders: respHeaders,
          responseBodyDecrypted: response.body,
          durationMs: sw.elapsedMilliseconds,
          error: 'HTTP ${response.statusCode}: ${response.body}',
        );
      }

      final (respEncrypted, decrypted) = await _decryptResponse(
        response.body,
        respHeaders,
        session,
      );

      sw.stop();
      return StepResult(
        step: stepNum,
        name: stepName,
        requestHeaders: headers,
        requestBodyPlaintext: plaintext,
        requestBodyEncrypted: encryptedPayload,
        responseHeaders: respHeaders,
        responseBodyEncrypted: respEncrypted,
        responseBodyDecrypted: decrypted,
        durationMs: sw.elapsedMilliseconds,
        success: true,
      );
    } catch (e) {
      sw.stop();
      return StepResult(
        step: stepNum,
        name: stepName,
        requestBodyPlaintext: plaintext,
        durationMs: sw.elapsedMilliseconds,
        error: e.toString(),
      );
    }
  }

  Future<(String, String)> _decryptResponse(
    String responseBody,
    Map<String, String> respHeaders,
    SessionContext session,
  ) async {
    final respKid = respHeaders['x-kid'];
    final respIdempKey = respHeaders['x-idempotency-key'];

    if (respKid != null && respIdempKey != null) {
      final parts = respIdempKey.split('.');
      final respAad = _crypto.buildAad(
        parts[0],
        parts[1],
        respKid,
        session.clientId,
      );

      final resBody = EncryptedPayload.fromJson(
        jsonDecode(responseBody) as Map<String, dynamic>,
      );

      final decrypted = await _crypto.decrypt(
        resBody.payload,
        session.sessionKey,
        respAad,
      );

      return (resBody.payload, decrypted);
    }

    return ('', responseBody);
  }

  Map<String, String> _extractResponseHeaders(http.Response response) {
    final result = <String, String>{};
    for (final key in ['x-kid', 'x-idempotency-key', 'content-type']) {
      final value = response.headers[key];
      if (value != null) result[key] = value;
    }
    return result;
  }
}

File: dotnet-external/Services/IdentityService.cs
using System.Diagnostics;
using System.Text;
using System.Text.Json;
using IdentityExample.Models;

namespace IdentityExample.Services;

/// <summary>
/// Token operations for external client.
/// Step 2: HMAC-SHA256 (X-Signature) instead of Basic Auth.
/// Steps 3-6: Bearer + GCM (same as internal).
/// </summary>
public sealed class IdentityService
{
    private readonly CryptoService _crypto;
    private readonly HmacService _hmac;
    private readonly IHttpClientFactory _httpClientFactory;
    private readonly string _sidecarUrl;
    private readonly string _clientId;
    private readonly string _clientSecret;

    public IdentityService(CryptoService crypto, HmacService hmac,
                           IHttpClientFactory httpClientFactory, IConfiguration config)
    {
        _crypto = crypto;
        _hmac = hmac;
        _httpClientFactory = httpClientFactory;
        _sidecarUrl = config["Sidecar:Url"] ?? "http://localhost:8141";
        _clientId = config["Client:Id"] ?? "external-partner-test";
        _clientSecret = config["Client:Secret"] ?? "";
    }

    /// <summary>Step 2: Issue tokens with HMAC-SHA256 + GCM.</summary>
    public async Task<StepResult> IssueTokenAsync(SessionContext session, string plaintext)
    {
        var sw = Stopwatch.StartNew();
        try
        {
            var nonce = _crypto.GenerateNonce();
            var ts = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds().ToString();
            var aad = _crypto.BuildAad(ts, nonce, session.Kid, session.ClientId);
            var encrypted = _crypto.AesGcmEncrypt(session.SessionKey, Encoding.UTF8.GetBytes(plaintext), aad);
            var payload = _crypto.ToBase64(encrypted);

            // HMAC over PLAINTEXT body (before encryption)
            var signature = _hmac.ComputeSignature("POST", "/api/v1/token/issue",
                ts, nonce, plaintext, _clientSecret);

            var headers = new Dictionary<string, string>
            {
                ["Content-Type"] = "application/json",
                ["X-Kid"] = session.Kid,
                ["X-ClientId"] = session.ClientId,
                ["X-Idempotency-Key"] = $"{ts}.{nonce}",
                ["X-Signature"] = signature
            };

            var (respHeaders, respEncrypted, decrypted) = await SendEncrypted(
                "/api/v1/token/issue", payload, headers, session);

            sw.Stop();
            return new StepResult(2, "Token Issue (HMAC + GCM)", headers, plaintext, payload,
                respHeaders, respEncrypted, decrypted, sw.ElapsedMilliseconds, true, null);
        }
        catch (Exception ex)
        {
            sw.Stop();
            return new StepResult(2, "Token Issue (HMAC + GCM)",
                new Dictionary<string, string>(), plaintext, "",
                new Dictionary<string, string>(), "", "",
                sw.ElapsedMilliseconds, false, ex.Message);
        }
    }

    /// <summary>Step 3: Introspect token (Bearer + GCM).</summary>
    public Task<StepResult> IntrospectTokenAsync(SessionContext session, string plaintext, string accessToken)
        => PostBearerEncrypted("/api/v1/introspect", session, plaintext, accessToken,
            3, "Token Introspection (Bearer + GCM)");

    /// <summary>Step 5: Refresh tokens (Bearer + GCM).</summary>
    public Task<StepResult> RefreshTokenAsync(SessionContext session, string plaintext, string accessToken)
        => PostBearerEncrypted("/api/v1/token", session, plaintext, accessToken,
            5, "Token Refresh (Bearer + GCM)");

    /// <summary>Step 6: Revoke token (Bearer + GCM).</summary>
    public Task<StepResult> RevokeTokenAsync(SessionContext session, string plaintext, string accessToken)
        => PostBearerEncrypted("/api/v1/revoke", session, plaintext, accessToken,
            6, "Token Revocation (Bearer + GCM)");

    private async Task<StepResult> PostBearerEncrypted(
        string fullPath, SessionContext session, string plaintext,
        string accessToken, int stepNum, string stepName)
    {
        var sw = Stopwatch.StartNew();
        try
        {
            var nonce = _crypto.GenerateNonce();
            var ts = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds().ToString();
            var aad = _crypto.BuildAad(ts, nonce, session.Kid, session.ClientId);
            var encrypted = _crypto.AesGcmEncrypt(session.SessionKey, Encoding.UTF8.GetBytes(plaintext), aad);
            var payload = _crypto.ToBase64(encrypted);

            var headers = new Dictionary<string, string>
            {
                ["Content-Type"] = "application/json",
                ["Authorization"] = $"Bearer {accessToken}",
                ["X-Kid"] = session.Kid,
                ["X-Idempotency-Key"] = $"{ts}.{nonce}",
                ["X-ClientId"] = session.ClientId
            };

            var (respHeaders, respEncrypted, decrypted) = await SendEncrypted(
                fullPath, payload, headers, session);

            sw.Stop();
            return new StepResult(stepNum, stepName, headers, plaintext, payload,
                respHeaders, respEncrypted, decrypted, sw.ElapsedMilliseconds, true, null);
        }
        catch (Exception ex)
        {
            sw.Stop();
            return new StepResult(stepNum, stepName,
                new Dictionary<string, string>(), plaintext, "",
                new Dictionary<string, string>(), "", "",
                sw.ElapsedMilliseconds, false, ex.Message);
        }
    }

    private async Task<(Dictionary<string, string> respHeaders, string respEncrypted, string decrypted)>
        SendEncrypted(string fullPath, string payload, Dictionary<string, string> headers, SessionContext session)
    {
        var requestBody = JsonSerializer.Serialize(new EncryptedPayload(payload));
        using var client = _httpClientFactory.CreateClient("sidecar");
        using var request = new HttpRequestMessage(HttpMethod.Post, $"{_sidecarUrl}{fullPath}");
        request.Content = new StringContent(requestBody, Encoding.UTF8, "application/json");
        foreach (var (k, v) in headers)
        {
            if (k.Equals("Content-Type", StringComparison.OrdinalIgnoreCase)) continue;
            request.Headers.TryAddWithoutValidation(k, v);
        }

        var response = await client.SendAsync(request);
        var responseBodyStr = await response.Content.ReadAsStringAsync();

        var respHeaders = new Dictionary<string, string>();
        foreach (var key in new[] { "x-kid", "x-idempotency-key", "content-type" })
        {
            if (response.Headers.TryGetValues(key, out var vals))
                respHeaders[key] = vals.First();
            else if (response.Content.Headers.TryGetValues(key, out var cVals))
                respHeaders[key] = cVals.First();
        }

        if (!response.IsSuccessStatusCode)
            throw new Exception($"HTTP {(int)response.StatusCode}: {responseBodyStr}");

        // Decrypt response
        var respKid = respHeaders.GetValueOrDefault("x-kid");
        var respIdempKey = respHeaders.GetValueOrDefault("x-idempotency-key");

        if (respKid != null && respIdempKey != null)
        {
            var parts = respIdempKey.Split('.');
            var respAad = _crypto.BuildAad(parts[0], parts[1], respKid, session.ClientId);
            var resBody = JsonSerializer.Deserialize<EncryptedPayload>(responseBodyStr)!;
            var decryptedBytes = _crypto.AesGcmDecrypt(session.SessionKey, respAad, _crypto.FromBase64(resBody.Payload));
            return (respHeaders, resBody.Payload, Encoding.UTF8.GetString(decryptedBytes));
        }

        return (respHeaders, "", responseBodyStr);
    }
}
